### ISSUE TITLE
How to reproduce the results of dgl?

### ISSUE THREAD

**Issue by WenxiongLiao** [dmlc/dgl#6736]
## ❓ Questions and Help

```

def seed_everything(seed_value):
    torch.manual_seed(seed_value)
    dgl.random.seed(seed_value)
    dgl.seed(seed_value)
    
    if torch.cuda.is_available(): 
        torch.cuda.manual_seed(seed_value)
        torch.cuda.manual_seed_all(seed_value)
        torch.backends.cudnn.benchmark = False
        torch.backends.cudnn.deterministic = True


class GAT(nn.Module):
    def __init__(self, in_dim, hidden_dim, out_dim, num_heads):
        super(GAT, self).__init__()
        self.layers = nn.ModuleList()
        self.layers.append(GATConv(in_dim, hidden_dim, num_heads, residual=True))
        self.layers.append(GATConv(hidden_dim * num_heads, out_dim, 1, residual=True))

    def forward(self, g, features):
        for conv in self.layers:
            features = conv(g, features).flatten(1)
            features = torch.relu(features)
        return features

result = []
for i in range(2):
    seed_everything(2022)
    adj_matrix = torch.randint(0, 2, (1000, 1000))
    adj_matrix = adj_matrix + torch.eye(adj_matrix.shape[0])
    g_nx = nx.from_numpy_array(adj_matrix.numpy())
    graph = dgl.from_networkx(g_nx)
    graph = dgl.add_self_loop(graph)

    graph_list = [graph] * 50
    batched_graph = dgl.batch(graph_list)
    features = torch.randn(batched_graph.num_nodes(), 1)
    model = GAT(in_dim=1, hidden_dim=32, out_dim=32, num_heads=2)

    device = "cuda:0"
    model.to(device)
    out = model(batched_graph.to(device), features.to(device))
    result.append(out)

torch.set_printoptions(precision=10)
print(result[0][result[0]!=result[1]])
print(result[1][result[0]!=result[1]])

```

**When I execute the same code twice (even though I fixed the random seed), the output results are different twice**

The above output is

```
tensor([0.0525254495, 0.0337213613, 0.2210013568,  ..., 0.0451040678,
        0.0790061727, 0.2986189723], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([0.0525254458, 0.0337213576, 0.2210013270,  ..., 0.0451040827,
        0.0790061653, 0.2986190319], device='cuda:0', grad_fn=<IndexBackward0>)
```

**How can I solve the above problem?**

**Comment by anko-intel** [comment]
You can see https://pytorch.org/docs/stable/notes/randomness.html#reproducibility.
On CPU inserting into seed_everything():
```
dgl.seed(seed)
torch.manual_seed(seed)
```
is enough.

**Comment by WenxiongLiao** [comment]
> You can see https://pytorch.org/docs/stable/notes/randomness.html#reproducibility. On CPU inserting into seed_everything():
> 
> ```
> dgl.seed(seed)
> torch.manual_seed(seed)
> ```
> 
> is enough.

**How to reproduce the results in gpu mode?**

**Comment by jermainewang** [comment]
Related issue: https://github.com/dmlc/dgl/issues/5425 . It is not easy to have deterministic results for arbitrary models since some GPU kernels are inherently non-deterministic.

**Comment by TristonC** [comment]
NVIDIA DGL team will check this issue. Thanks @WenxiongLiao  for reporting. We can repro your issue.

**Comment by TristonC** [comment]
Our team has done some triage into this issue. Looks like that the with CUDA the float32 datatype was used, so the precision can go to about 7 digits. We recommend doing either 6- or 7-digits precision comparison.

**Comment by jermainewang** [comment]
Thanks @TristonC  for the investigation. Closing this.
